File: 17.RollingHashAlgo.py
# A Rolling Hash (or Polynomial Hash) is a technique to efficiently 
# compute hash values of substrings by reusing previous computations.
# It’s widely used in string matching algorithms like Rabin-Karp to speed up pattern searching.

# Explanation

# Normally, calculating a hash for each substring takes O(n⋅m) time 
# (where n is text length and m is substring length).

# With Rolling Hash, we reuse the previous hash value to calculate 
# the next one in constant time O(1).

# It treats the string like a number in a base system (like base p), 
# with modulus M to avoid overflow.

# Formula for a string s = s[0]s[1]...s[n-1] is:

# H(s) = (s[0] ⋅ p^(n−1) + s[1] ⋅ p^(n−2) + ... + s[n−1]) mod M

# When sliding over text, we "remove" the first character and "add" 
# the next character to update the hash efficiently.

p = 31
M = 10**9 + 9

def compute_hash(s):
    h = 0
    p_pow = 1
    for ch in s:

        h = (h + (ord(ch) - 96) * p_pow) % M
        p_pow = (p_pow * p) % M
    return h

text = "abacaba"
pattern = "aba"

pattern_hash = compute_hash(pattern)

n = len(text)
prefix_hash = [0] * (n + 1)
p_pow = [1] * (n + 1)

for i in range(1, n + 1):
    prefix_hash[i] = (prefix_hash[i - 1] + (ord(text[i - 1]) - 96) * p_pow[i - 1]) % M
    p_pow[i] = (p_pow[i - 1] * p) % M

matches = []
for i in range(n - len(pattern) + 1):
    cur_hash = (prefix_hash[i + len(pattern)] - prefix_hash[i] + M) % M
    if cur_hash == (pattern_hash * p_pow[i]) % M:
        matches.append(i)

print("Pattern found at indices:", matches)

File: 17.RollingHashAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;

public class RollingHashExample {
    static final int p = 31;                // base for characters (like size of alphabet)
    static final int M = 1000000009;        // large prime modulus

    // Function to compute hash of a string
    static long computeHash(String s) {
        long h = 0;
        long p_pow = 1;

        for (int i = 0; i < s.length(); i++) {
            // (s.charAt(i) - 'a' + 1) converts 'a'->1, 'b'->2, ...
            h = (h + (s.charAt(i) - 'a' + 1) * p_pow) % M;
            p_pow = (p_pow * p) % M;
        }
        return h;
    }

    public static void main(String[] args) throws Exception {
        String text = "abacaba";
        String pattern = "aba";

        // Step 1: compute hash of pattern
        long patternHash = computeHash(pattern);

        // Step 2: precompute prefix hashes for text
        int n = text.length();
        long[] prefixHash = new long[n + 1];
        long[] p_pow = new long[n + 1];
        p_pow[0] = 1;

        for (int i = 1; i <= n; i++) {
            prefixHash[i] = (prefixHash[i - 1] + (text.charAt(i - 1) - 'a' + 1) * p_pow[i - 1]) % M;
            p_pow[i] = (p_pow[i - 1] * p) % M;
        }

        // Step 3: search pattern in text using rolling hash
        System.out.print("Pattern found at indices: ");
        for (int i = 0; i <= n - pattern.length(); i++) {
            long curHash = (prefixHash[i + pattern.length()] - prefixHash[i] + M) % M;

            if (curHash == (patternHash * p_pow[i]) % M) {
                System.out.print(i + " ");
            }
        }
    }
}
